cub/cub/device/device_merge.cuh

```diff
@@ -0,0 +1,195 @@
+  //! which means any two equivalent values (neither value is ordered before the other) may be written to the ouput
+  //! sequence in any order.
+  //!
+  //! @par Snippet
+  //! The code snippet below illustrates the merging of two device vectors of `int` keys.
+  //! @par
+  //! @code
+  //! .. literalinclude:: ../../test/catch2_test_device_merge_api.cu
+  //!     :language: c++
+  //!     :dedent:
+  //!     :start-after: example-begin merge-keys
+  //!     :end-before: example-end merge-keys
+  //! @endcode
+  //!
+  //! @tparam KeyIteratorIn1 **[deduced]** Random access iterator to the first sorted input sequence. Must have the same
+  //! value type as KeyIteratorIn2.
+  //! @tparam KeyIteratorIn2 **[deduced]** Random access iterator to the second sorted input sequence. Must have the
+  //! same value type as KeyIteratorIn1.
+  //! @tparam KeyIteratorOut **[deduced]** Random access iterator to the output sequence.
+  //! @tparam Offset **[deduced]** Integral used for size and offset calculations.
+  //! @tparam CompareOp **[deduced]** Binary predicate to compare the input iterator's value types. Must have a
+  //! signature equivalent to `bool operator()(Key lhs, Key rhs)` and establish a [strict weak ordering].
+  //!
+  //! @param[in] d_temp_storage Device-accessible allocation of temporary storage. When `nullptr`, the required
+  //! allocation size is written to `temp_storage_bytes` and no work is done.
+  //! @param[in,out] temp_storage_bytes Reference to size in bytes of `d_temp_storage` allocation.
+  //! @param[in] keys_in1 Iterator to the beginning of the first sorted input sequence.
+  //! @param[in] num_keys1 Number of keys in the first input sequence.
+  //! @param[in] keys_in2 Iterator to the beginning of the second sorted input sequence.
+  //! @param[in] num_keys2 Number of keys in the second input sequence.
+  //! @param[in] keys_out Iterator to the beginning of the output sequence.
+  //! @param[in] compare_op Comparison function object, returning true if the first argument is ordered before the
+  //! second. Must establish a [strict weak ordering].
+  //! @param[in] stream **[optional]** CUDA stream to launch kernels into. Default is stream<sub>0</sub>.
+  //!
+  //! [strict weak ordering]: https://en.cppreference.com/w/cpp/concepts/strict_weak_order
+  template <typename KeyIteratorIn1,
+            typename KeyIteratorIn2,
+            typename KeyIteratorOut,
+            typename Offset,
+            typename CompareOp = ::cuda::std::less<>>
+  CUB_RUNTIME_FUNCTION static cudaError_t MergeKeys(
+    void* d_temp_storage,
+    std::size_t& temp_storage_bytes,
+    KeyIteratorIn1 keys_in1,
+    Offset num_keys1,
+    KeyIteratorIn2 keys_in2,
+    Offset num_keys2,
+    KeyIteratorOut keys_out,
+    CompareOp compare_op = {},
+    cudaStream_t stream  = 0)
+  {
+    // TODO(georgii): hardcode Offset to int for now
```

Comment:
@gevtushenko If I hardcoded the offset type to `int` in the CUB public API, but leave it flexible in the dispatch layer, I cannot use the launch wrappers anymore in my unit tests, which is super inconvenient. Can I just leave the CUB API offset type as a template?

```suggestion
  template <typename KeyIteratorIn1,
            typename KeyIteratorIn2,
            typename KeyIteratorOut,
            typename CompareOp = ::cuda::std::less<>>
  CUB_RUNTIME_FUNCTION static cudaError_t MergeKeys(
    void* d_temp_storage,
    std::size_t& temp_storage_bytes,
    KeyIteratorIn1 keys_in1,
    int num_keys1,
    KeyIteratorIn2 keys_in2,
    int num_keys2,
    KeyIteratorOut keys_out,
    CompareOp compare_op = {},
    cudaStream_t stream  = 0)
  {
    // TODO(georgii): hardcode Offset to int for now
```

Reply:
Unfortunately, we'll have to go with `int` offsets for device API. Allowing arbitrary types would put significant maintenance load on us (testing, benchmarking, tuning etc). On the test end, you can use wrappers of integer types and following something along this https://github.com/NVIDIA/cccl/pull/1830 for large offset types.